File: backend/src/bajo_nivel/buffers/buf.c
#include <stdio.h>
#include <string.h>   // Aquí se incluyen las funciones de memoria
#include <stdlib.h>

/* Aclarar que el buffer no es un concepto exclusivo de C, ni hay
    ninguna forma de implemetarlo explícita. Reservar un espacio
    en memoria e ir iterando por los bloques significa manipular
    un buffer.
*/

int main() {
    // =======================================================
    //                       memset()
    /* Se usa para inicializar un número de bytes específico
        a un valor determinado dentro del buffer.
       Toma 3 parámetros:
           - Puntero a un área de memoria: el área de memoria a donde
               apunte el puntero pasado como primer parámetro será donde
               se establezcan los bytes.
           - Valor de los elementos: el segundo parámetro indica el valor
               que tomarán todos los elementos dentro de ese área de memoria.
           - Tamaño del área de memoria: el tercer parámetro requiere especificar
               el tamañ de memoria del área donde estamos trabajando.
    */
    int i;

    // Asignamos memoria para un array de 5 elementos
    char *a = (char *) malloc(5*sizeof(char));

    printf("Valores del buffer antes de ejecutar memset()\n");
    for (i = 0; i < 5; ++i)
        printf("  a[%d] = %d  ", i,    a[i]);

    // Establecemos el valor de todos los elementos del buffer a 3
    memset(a, 3, 5*sizeof(char));
    printf("\n\nValores después de ejecutar memset()\n");

    for (i = 0; i < 5; ++i)
        printf("  a[%d] = %d  ", i,    a[i]);
    printf("\n\n");

    // Liberamos el espacio asignado de la memoria
    free(a);

    // ======================================================

    printf("===============================================\n");

    // ======================================================
    //                     memcpy()
    /* Se utiliza para copiar una cantidad específica de bytes
        de un área de memoria a otra */
    char saludo1[5] = "hola";
    char saludo2[5];

    if ( memcpy(saludo2, saludo1, strlen(saludo1)) ) {
        printf("Los elementos de saludo1 han sido copiados a saludo2\n");
        printf("saludo1 = %s\nsaludo2 = %s\n", saludo1, saludo2);
    } else {
        printf("Hubo un error copiando saludo1 en saludo2\n");
    }

    // ======================================================

    return 0;
}

File: backend/src/bajo_nivel/buffers/buf.go
package main

/* Un objeto 'bytes.Buffer' es un buffer de bytes
    de tamaño variable con los métodos 'Read' y 'Write'.
    El valor 0 para 'Buffer' es un buffer vacío listo
    para usar.
*/

import (
    "fmt"
    "os"
    "bytes"
)

func main() {
    // Creamos un buffer vacío
    var b bytes.Buffer    // Un buffer no necesita inicialización

    // Escribir en el buffer
    b.Write([]byte("¡Hola "))
    // https://golang.org/pkg/bytes/#Buffer.Write

    // Cambiar el contenido del buffer
    fmt.Fprintf(&b, "mundo!\n")  // La función 'fmt.Fprintf' escribe
                               //  la salida en el buffer del primer parámetro
                               //  https://golang.org/pkg/fmt/#Fprintf

    // Imprimir el contenido del buffer
    b.WriteTo(os.Stdout)  // ¡Hola mundo!
    // https://golang.org/pkg/bytes/#Buffer.WriteTo

}

/* Fuentes:
https://golang.org/pkg/bytes/#example_Buffer
*/
